In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display all rows and columns of a dataframe instead of a truncated version
from IPython.display import display
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [5]:
with open('../Data/Feature_engr/Ames_Housing_Feature_Description.txt','r') as f: 
    print(f.read())

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

In [3]:
df = pd.read_csv('../Data/Feature_engr/Ames_NO_Missing_Data.csv')
df.head()

MS SubClass MS Zoning  Lot Frontage  Lot Area Street Lot Shape  \
0           20        RL         141.0     31770   Pave       IR1   
1           20        RH          80.0     11622   Pave       Reg   
2           20        RL          81.0     14267   Pave       IR1   
3           20        RL          93.0     11160   Pave       Reg   
4           60        RL          74.0     13830   Pave       IR1   

  Land Contour Utilities Lot Config Land Slope Neighborhood Condition 1  \
0          Lvl    AllPub     Corner        Gtl        NAmes        Norm   
1          Lvl    AllPub     Inside        Gtl        NAmes       Feedr   
2          Lvl    AllPub     Corner        Gtl        NAmes        Norm   
3          Lvl    AllPub     Corner        Gtl        NAmes        Norm   
4          Lvl    AllPub     Inside        Gtl      Gilbert        Norm   

  Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0        Norm      1Fam      1Story             6             5        1960   
1        Norm      1Fam      1Story             5             6        1961   
2        Norm      1Fam      1Story             6             6        1958   
3        Norm      1Fam      1Story             7             5        1968   
4        Norm      1Fam      2Story             5             5        1997   

   Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type  \
0            1960        Hip   CompShg      BrkFace      Plywood        Stone   
1            1961      Gable   CompShg      VinylSd      VinylSd         None   
2            1958        Hip   CompShg      Wd Sdng      Wd Sdng      BrkFace   
3            1968        Hip   CompShg      BrkFace      BrkFace         None   
4            1998      Gable   CompShg      VinylSd      VinylSd         None   

   Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual Bsmt Cond  \
0         112.0         TA         TA     CBlock        TA        Gd   
1           0.0         TA         TA     CBlock        TA        TA   
2         108.0         TA         TA     CBlock        TA        TA   
3           0.0         Gd         TA     CBlock        TA        TA   
4           0.0         TA         TA      PConc        Gd        TA   

  Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  \
0            Gd            BLQ         639.0            Unf           0.0   
1            No            Rec         468.0            LwQ         144.0   
2            No            ALQ         923.0            Unf           0.0   
3            No            ALQ        1065.0            Unf           0.0   
4            No            GLQ         791.0            Unf           0.0   

   Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air Electrical  \
0        441.0         1080.0    GasA         Fa           Y      SBrkr   
1        270.0          882.0    GasA         TA           Y      SBrkr   
2        406.0         1329.0    GasA         TA           Y      SBrkr   
3       1045.0         2110.0    GasA         Ex           Y      SBrkr   
4        137.0          928.0    GasA         Gd           Y      SBrkr   

   1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  \
0        1656           0                0         1656             1.0   
1         896           0                0          896             0.0   
2        1329           0                0         1329             0.0   
3        2110           0                0         2110             1.0   
4         928         701                0         1629             0.0   

   Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  Kitchen AbvGr  \
0             0.0          1          0              3              1   
1             0.0          1          0              2              1   
2             0.0          1          1              3              1   
3             0.0          2          1              3              1   
4           

In [4]:
df.isnull().sum()

MS SubClass        0
MS Zoning          0
Lot Frontage       0
Lot Area           0
Street             0
Lot Shape          0
Land Contour       0
Utilities          0
Lot Config         0
Land Slope         0
Neighborhood       0
Condition 1        0
Condition 2        0
Bldg Type          0
House Style        0
Overall Qual       0
Overall Cond       0
Year Built         0
Year Remod/Add     0
Roof Style         0
Roof Matl          0
Exterior 1st       0
Exterior 2nd       0
Mas Vnr Type       0
Mas Vnr Area       0
Exter Qual         0
Exter Cond         0
Foundation         0
Bsmt Qual          0
Bsmt Cond          0
Bsmt Exposure      0
BsmtFin Type 1     0
BsmtFin SF 1       0
BsmtFin Type 2     0
BsmtFin SF 2       0
Bsmt Unf SF        0
Total Bsmt SF      0
Heating            0
Heating QC         0
Central Air        0
Electrical         0
1st Flr SF         0
2nd Flr SF         0
Low Qual Fin SF    0
Gr Liv Area        0
Bsmt Full Bath     0
Bsmt Half Bath     0
Full Bath    

## Numerical Column to Categorical

We need to be careful when it comes to encoding categories as numbers. We want to make sure that the numerical relationship makes sense for a model. For example, the encoding MSSubClass is essentially just a number code per class:

    MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

The number itself does not appear to have a relationship to the other numbers. While 30 > 20 is True, it doesn't really make sense that "1-STORY 1945 & OLDER" > "1-STORY 1946 & NEWER ALL STYLES". Keep in mind, this isn't always the case, for example 1st class seats versus 2nd class seats encoded as 1 and 2. Make sure you fully understand your data set to examine what needs to be converted/changed.

### MSSubClass

In [6]:
# Convert to String
df['MS SubClass'] = df['MS SubClass'].apply(str)

## Creating "Dummy" Variables

In [7]:
# Avoiding MultiCollinearity and the Dummy Variable Trap
person_state =  pd.Series(['Dead','Alive','Dead','Alive','Dead','Dead'])
person_state

0     Dead
1    Alive
2     Dead
3    Alive
4     Dead
5     Dead
dtype: object

In [8]:
pd.get_dummies(person_state)

Alive  Dead
0      0     1
1      1     0
2      0     1
3      1     0
4      0     1
5      0     1

In [9]:
pd.get_dummies(person_state,drop_first=True)

Dead
0     1
1     0
2     1
3     0
4     1
5     1

### Creating Dummy Variables from Object Columns

In [11]:
# All columns that are strings/object
df.select_dtypes(include='object').head()

MS SubClass MS Zoning Street Lot Shape Land Contour Utilities Lot Config  \
0          20        RL   Pave       IR1          Lvl    AllPub     Corner   
1          20        RH   Pave       Reg          Lvl    AllPub     Inside   
2          20        RL   Pave       IR1          Lvl    AllPub     Corner   
3          20        RL   Pave       Reg          Lvl    AllPub     Corner   
4          60        RL   Pave       IR1          Lvl    AllPub     Inside   

  Land Slope Neighborhood Condition 1 Condition 2 Bldg Type House Style  \
0        Gtl        NAmes        Norm        Norm      1Fam      1Story   
1        Gtl        NAmes       Feedr        Norm      1Fam      1Story   
2        Gtl        NAmes        Norm        Norm      1Fam      1Story   
3        Gtl        NAmes        Norm        Norm      1Fam      1Story   
4        Gtl      Gilbert        Norm        Norm      1Fam      2Story   

  Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type Exter Qual  \
0        Hip   CompShg      BrkFace      Plywood        Stone         TA   
1      Gable   CompShg      VinylSd      VinylSd         None         TA   
2        Hip   CompShg      Wd Sdng      Wd Sdng      BrkFace         TA   
3        Hip   CompShg      BrkFace      BrkFace         None         Gd   
4      Gable   CompShg      VinylSd      VinylSd         None         TA   

  Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
0         TA     CBlock        TA        Gd            Gd            BLQ   
1         TA     CBlock        TA        TA            No            Rec   
2         TA     CBlock        TA        TA            No            ALQ   
3         TA     CBlock        TA        TA            No            ALQ   
4         TA      PConc        Gd        TA            No            GLQ   

  BsmtFin Type 2 Heating Heating QC Central Air Electrical Kitchen Qual  \
0            Unf    GasA         Fa           Y      SBrkr           TA   
1            LwQ    GasA         TA           Y      SBrkr           TA   
2            Unf    GasA         TA           Y      SBrkr           Gd   
3            Unf    GasA         Ex           Y      SBrkr           Ex   
4            Unf    GasA         Gd           Y      SBrkr           TA   

  Functional Fireplace Qu Garage Type Garage Finish Garage Qual Garage Cond  \
0        Typ           Gd      Attchd           Fin          TA          TA   
1        Typ         None      Attchd           Unf          TA          TA   
2        Typ         None      Attchd           Unf          TA          TA   
3        Typ           TA      Attchd           Fin          TA          TA   
4        Typ           TA      Attchd           Fin          TA          TA   

  Paved Drive Sale Type Sale Condition  
0           P       WD          Normal  
1           Y       WD          Normal  
2           Y       WD          Normal  
3           Y       WD          Normal  
4           Y       WD          Normal

In [13]:
# df.info()

In [14]:
my_object_df = df.select_dtypes(include='object')
my_numeric_df = df.select_dtypes(exclude='object')
df_object_dummies = pd.get_dummies(my_object_df, drop_first=True)

In [16]:
df_object_dummies.shape

(2925, 238)

In [17]:
final_df = pd.concat([my_numeric_df, df_object_dummies], axis=1)
final_df.head()

Lot Frontage  Lot Area  Overall Qual  Overall Cond  Year Built  \
0         141.0     31770             6             5        1960   
1          80.0     11622             5             6        1961   
2          81.0     14267             6             6        1958   
3          93.0     11160             7             5        1968   
4          74.0     13830             5             5        1997   

   Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  \
0            1960         112.0         639.0           0.0        441.0   
1            1961           0.0         468.0         144.0        270.0   
2            1958         108.0         923.0           0.0        406.0   
3            1968           0.0        1065.0           0.0       1045.0   
4            1998           0.0         791.0           0.0        137.0   

   Total Bsmt SF  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0         1080.0        1656           0                0         1656   
1          882.0         896           0                0          896   
2         1329.0        1329           0                0         1329   
3         2110.0        2110           0                0         2110   
4          928.0         928         701                0         1629   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0             1.0             0.0          1          0              3   
1             0.0             0.0          1          0              2   
2             0.0             0.0          1          1              3   
3             1.0             0.0          2          1              3   
4             0.0             0.0          2          1              3   

   Kitchen AbvGr  TotRms AbvGrd  Fireplaces  Garage Yr Blt  Garage Cars  \
0              1              7           2         1960.0          2.0   
1              1              5           0         1961.0          1.0   
2              1              6           0         1958.0          1.0   
3              1              8           2         1968.0          2.0   
4              1              6           1         1997.0          2.0   

   Garage Area  Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  \
0        528.0           210             62               0           0   
1        730.0           140              0               0           0   
2        312.0           393             36               0           0   
3        522.0             0              0               0           0   
4        482.0           212             34               0           0   

   Screen Porch  Pool Area  Misc Val  Mo Sold  Yr Sold  SalePrice  \
0             0          0         0        5     2010     215000   
1           120          0         0        6     2010     105000   
2             0          0     12500        6     2010     172000   
3             0          0         0        4     2010     244000   
4             0          0         0        3     2010     189900   

   MS SubClass_150  MS SubClass_160  MS SubClass_180  MS SubClass_190  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   MS SubClass_20  MS SubClass_30  MS SubClass_40  MS SubClass_45  \
0               1               0               0               0   
1               1               0               0               0   
2               1               0               0               0   
3               1               0               0               0   
4               0               0               0               0   

   MS SubClass_50  MS SubClass_60  MS SubClass_70  MS SubClass_75  \
0

In [18]:
final_df.shape

(2925, 274)

In [19]:
final_df.corr()['SalePrice'].sort_values()

Exter Qual_TA            -0.591459
Kitchen Qual_TA          -0.527461
Fireplace Qu_None        -0.481740
Bsmt Qual_TA             -0.453022
Garage Finish_Unf        -0.422363
Mas Vnr Type_None        -0.395083
Garage Type_Detchd       -0.365209
Foundation_CBlock        -0.350087
Heating QC_TA            -0.338812
Lot Shape_Reg            -0.300738
MS Zoning_RM             -0.292661
Bsmt Exposure_No         -0.283246
Roof Style_Gable         -0.256806
MS SubClass_30           -0.234964
Garage Type_None         -0.226028
Garage Cond_None         -0.225746
Garage Finish_None       -0.225746
Garage Qual_None         -0.225746
Neighborhood_OldTown     -0.212042
Sale Type_WD             -0.205144
Neighborhood_NAmes       -0.188772
MS SubClass_50           -0.179005
Neighborhood_IDOTRR      -0.174880
Exterior 1st_Wd Sdng     -0.173195
Exterior 2nd_Wd Sdng     -0.167913
Neighborhood_Edwards     -0.167587
Garage Qual_Fa           -0.165818
BsmtFin Type 1_Rec       -0.157269
Bsmt Qual_Fa        

In [20]:
# final_df.to_csv('../Data/Feature_engr/AMES_Final_DF.csv')